### ISSUE TITLE
Optimizations

### ISSUE THREAD

**Issue by NikoOinonen** [Probe-Particle/ppafm#84]

**Comment by ProkopHapala** [comment]
> The Lennard-Jones calculation could be optimized by cutting out atoms beyond some cut-off distance out of the force-field box. The LJ force fades very quickly so especially with periodic boundaries we have a lot of completely redundant atoms that don't contribute to the force in any way.

I don't think it is worth it. If you calculate electrostatics for given atom, then evaluation vdW adds only only minor additional cost, because main cost on GPU is global memory I/O (read atomic position and parameters, store output to texture). This is the reason why I calculate vdW and electrostatics in one kernel.

What coulb be better way to optimize this:
*  put loop over PBC cells inside OpenCL kernell (now we pass array of multiplied coordinates), so the kernel reads parameters/positions only for atoms in original (not multiplied) cell. 
* Use multiplole expansion for electrostatics of distant PBC cells beyond nearest neighbor.  (but that makes sense only for small cells otherwise we use only nearest neighbor replica)

**Comment by NikoOinonen** [comment]
> I don't think it is worth it. If you calculate electrostatics for given atom, then evaluation vdW adds only only minor additional cost, because main cost on GPU is global memory I/O (read atomic position and parameters, store output to texture). This is the reason why I calculate vdW and electrostatics in one kernel.

I see your point, but I was originally thinking that this would only apply to the case where the electrostatics are computed separately, i.e. FFT with the Hartree potential, or when the electrostatics are absent. When you have a large periodic system with a surface slab, the LJ calculation time is actually significant, often longer than the electrostatics FFT. I was thinking that if you would more carefully consider which atoms are actually inside or close to the FF box, then you could eliminate a lot of irrelevant calculations, and you can do this pruning of atoms before uploading the atom parameters to the device.
